arangod/Pregel/Algos/WCC.cpp

```diff
@@ -40,95 +40,136 @@ using namespace arangodb::pregel::algos;
+   * kicks of the algorithm and does not yet have input.
+   */
+  bool selectMinimumOfLocalAndInput(MessageIterator<SenderMessage<uint64_t>> const& messages) {
+    // On first iteration, we need to propagate.
+    // Otherwise the default is to stay silent, unless some message
+    // sends a different component then us.
+    // Either the sender has a wrong component or we have.
+    bool shouldPropagate = globalSuperstep() == 0;
+
+    auto& myData = *mutableVertexData();
+    for (const SenderMessage<uint64_t>* msg : messages) {
+      if (globalSuperstep() == 1) {
+        // In the first step, we need to retain all inbound connections
+        // for propagation
+        myData.inboundNeighbors.emplace(msg->senderId);
       }
-      
-      SenderMessage<uint64_t> message(pregelId(), currentComponent);
-      for (const SenderMessage<uint64_t>* msg : messages) {
-        if (msg->value > currentComponent) {
-          TRI_ASSERT(msg->senderId != pregelId());
-          sendMessage(msg->senderId, message);
-          halt = false;
+      if (msg->value != myData.component) {
+        // we have a difference. Send updates
+        shouldPropagate = true;
+        if (msg->value < myData.component) {
+          // The other component is lower.
+          // We join this component
+          myData.component = msg->value;
         }
       }
-
-      if (currentComponent != vertexData()) {
-        *mutableVertexData() = currentComponent;
-        halt = false;
-      }
-      
-      if (halt) {
-        voteHalt();
-      } else {
-        voteActive();
-      }
     }
-       
-    if (this->getEdgeCount() > 0) {
-      SenderMessage<uint64_t> message(pregelId(), currentComponent);    
-      RangeIterator<Edge<uint64_t>> edges = this->getEdges();
-      for (; edges.hasMore(); ++edges) {
-        Edge<uint64_t>* edge = *edges;
-        if (edge->toKey() == this->key()) {
-          continue; // no need to send message to self
-        }
-      
-        // remember the value we send
-        edge->data() = currentComponent;
+    return shouldPropagate;
+  }
 
-        sendMessage(edge, message);
+  /**
+   * @brief
+   * Send the current vertex data to all our neighbors, inbound
+   * and outbound.
+   * Store the component value in the outbound edges
+   */
+  void propagate() {
+    auto const& myData = vertexData();
+    SenderMessage<uint64_t> message(pregelId(), myData.component);
+    // Send to OUTBOUND neighbors
+    RangeIterator<Edge<uint64_t>> edges = this->getEdges();
+    for (; edges.hasMore(); ++edges) {
+      Edge<uint64_t>* edge = *edges;
+      if (edge->toKey() == this->key()) {
+        continue;  // no need to send message to self
       }
+
+      // remember the value we send
+      // NOTE: I have done refactroing of the algorithm
+      // the original variant saved this, i do not know
+      // if it is actually relevant for anything.
+      edge->data() = myData.component;
```

Comment:
I think you can remove it. It is never accessed and Pregel can not write data into edges, I don't see what this should be good for.